Plutarch/NonZero.hs

```diff
@@ -0,0 +1,77 @@
+{-# LANGUAGE FlexibleInstances #-}
+{-# LANGUAGE UndecidableInstances #-}
+
+module Plutarch.NonZero (PNonZero, pnonZero, ptryNonZero) where
+
+import Data.Functor.Const (Const)
+import GHC.Generics (Generic)
+
+import Plutarch.Bool (PEq, POrd, pif, (#==))
+import Plutarch.Builtin (PAsData, PData, PIsData, pdata)
+import Plutarch.Integer (PInteger, PIntegral)
+
+import Plutarch.Maybe (PMaybe (PJust, PNothing))
+
+import Plutarch (
+  DerivePlutusType (DPTStrat),
+  PlutusType,
+  PlutusTypeNewtype,
+  Term,
+  TermCont (runTermCont),
+  pcon,
+  phoistAcyclic,
+  plam,
+  plet,
+  pto,
+  (#),
+  (#$),
+  type (:-->),
+ )
+import Plutarch.Num (PNum (pfromInteger, (#-)))
+import Plutarch.Show (PShow)
+import Plutarch.TermCont (tcont)
+import Plutarch.Trace (ptraceError)
+import Plutarch.TryFrom (PTryFrom (PTryFromExcess, ptryFrom'), ptryFrom)
+
+newtype PNonZero s = PNonZero (Term s PInteger)
+  deriving stock (Generic)
+  deriving anyclass (PlutusType, PIsData, PEq, POrd, PIntegral, PShow)
+instance DerivePlutusType PNonZero where type DPTStrat _ = PlutusTypeNewtype
+
+instance PNum PNonZero where
+  x #- y = ptryNonZero #$ pto x #- pto y
+
+  pfromInteger 0 = error "pnonZero.pfromInteger: encountered 0"
```

Comment:
This currently errors at Plutarch compile time - I feel like there should be a way to produce `Term`s that are just descriptive `CompilerError`s (raised by `compile` - eventually).

Reply:
pthrow

Reply:
updated